### PR TITLE
Arnej/optimize copy cells

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@baldersheim as the hash_map / optimization expert please review
@geirst  and @toregge - feel free to veto the second commit if you think it's too "clever"

### CODE REVIEW

**Review by toregge — APPROVED** [review]

File: eval/src/vespa/eval/tensor/sparse/sparse_tensor.cpp [hunk 1]
```diff
@@ -29,10 +29,11 @@ using Cells = SparseTensor::Cells;
 void
 copyCells(Cells &cells, const Cells &cells_in, Stash &stash)
 {
-    for (const auto &cell : cells_in) {
+    cells = cells_in;
+    for (auto &cell : cells) {
         SparseTensorAddressRef oldRef = cell.first;
         SparseTensorAddressRef newRef(oldRef, stash);
-        cells[newRef] = cell.second;
+        cell.first = newRef;
```
Comment (baldersheim) [inline eval/src/vespa/eval/tensor/sparse/sparse_tensor.cpp:36]: 1 - A little comment about this not changing the hash value would be good.
2 - I guess you can can also write it as 
cell.first = SparseTensorAddressRef(cell.first, stash);
You should see what effect it has on speed. The compiler might find it easier or harder. 
noexcept on the SparseTensorAddressRef constructors might also make it easier on the optimizer.

**Review by baldersheim — CHANGES_REQUESTED** [review]
At least a comment about dirty trick would be good.

File: eval/src/vespa/eval/tensor/sparse/sparse_tensor.cpp (hunk 1 above)
Comment (arnej27959) [reply eval/src/vespa/eval/tensor/sparse/sparse_tensor.cpp:40]: added some comments.
doing 2) or adding noexcept didn't have any impact on the produced code, so I left it as-is.

**Comment by arnej27959** [comment]
@baldersheim Please Take Another Look.

**Review by baldersheim — APPROVED** [review]

**Review by baldersheim — APPROVED** [review]

**Pull request merged**